opttoy2/opttoy.go

```diff
@@ -0,0 +1,1165 @@
+	isNotOp:             scalar,
+	anyOp:               scalar,
+	someOp:              scalar,
+	allOp:               scalar,
+	bitandOp:            scalar,
+	bitorOp:             scalar,
+	bitxorOp:            scalar,
+	plusOp:              scalar,
+	minusOp:             scalar,
+	multOp:              scalar,
+	divOp:               scalar,
+	floorDivOp:          scalar,
+	modOp:               scalar,
+	powOp:               scalar,
+	concatOp:            scalar,
+	lShiftOp:            scalar,
+	rShiftOp:            scalar,
+	unaryPlusOp:         scalar,
+	unaryMinusOp:        scalar,
+	unaryComplementOp:   scalar,
+}
+
+var comparisonOpMap = [...]operator{
+	parser.EQ:                eqOp,
+	parser.LT:                ltOp,
+	parser.GT:                gtOp,
+	parser.LE:                leOp,
+	parser.GE:                geOp,
+	parser.NE:                neOp,
+	parser.In:                inOp,
+	parser.NotIn:             notInOp,
+	parser.Like:              likeOp,
+	parser.NotLike:           notLikeOp,
+	parser.ILike:             iLikeOp,
+	parser.NotILike:          notILikeOp,
+	parser.SimilarTo:         similarToOp,
+	parser.NotSimilarTo:      notSimilarToOp,
+	parser.RegMatch:          regMatchOp,
+	parser.NotRegMatch:       notRegMatchOp,
+	parser.RegIMatch:         regIMatchOp,
+	parser.NotRegIMatch:      notRegIMatchOp,
+	parser.IsDistinctFrom:    isDistinctFromOp,
+	parser.IsNotDistinctFrom: isNotDistinctFromOp,
+	parser.Is:                isOp,
+	parser.IsNot:             isNotOp,
+	parser.Any:               anyOp,
+	parser.Some:              someOp,
+	parser.All:               allOp,
+}
+
+var binaryOpMap = [...]operator{
+	parser.Bitand:   bitandOp,
+	parser.Bitor:    bitorOp,
+	parser.Bitxor:   bitxorOp,
+	parser.Plus:     plusOp,
+	parser.Minus:    minusOp,
+	parser.Mult:     multOp,
+	parser.Div:      divOp,
+	parser.FloorDiv: floorDivOp,
+	parser.Mod:      modOp,
+	parser.Pow:      powOp,
+	parser.Concat:   concatOp,
+	parser.LShift:   lShiftOp,
+	parser.RShift:   rShiftOp,
+}
+
+var unaryOpMap = [...]operator{
+	parser.UnaryPlus:       unaryPlusOp,
+	parser.UnaryMinus:      unaryMinusOp,
+	parser.UnaryComplement: unaryComplementOp,
+}
+
+// expr is a unified interface for both relational and scalar expressions in a
+// query. Expressions have optional inputs, projections and
+// filters. Additionally, an expression maintains a bitmap of required input
+// variables and a bitmap of the output variables it generates.
+//
+// Every unique column and every projection (that is more than just a pass
+// through of a variable) is given a variable index with the query. The
+// variable indexes are global to the query (see queryState) making the bitmaps
+// easily comparable. For example, consider the query:
+//
+//   SELECT x FROM a WHERE y > 0
+//
+// There are 2 variables in the above query: x and y. During name resolution,
+// the above query becomes:
+//
+//   SELECT @0 FROM a WHERE @1 > 0
+//   -- @0 -> x
+//   -- @1 -> y
```

Comment:
According to the code below I understand there should be a 3rd expression `@3 -> @1 > 0`

i.e. all predicates also get indexes in the bitmap, for otherwise we can't group them together.

Reply:
Currently in this code, predicates do not get indexes in the bitmap. A scalar expression only gets an index if it is used as a projection. I'm not following your concern about not being able to group them together. You've probably stumbled on to a case where the current code is inadequate.